pandas/io/parsers/base_parser.py

```diff
@@ -798,6 +799,26 @@ def _do_date_conversions(self, names, data):
 
         return names, data
 
+    def _check_data_length(self, columns: List[str], data: List[ArrayLike]):
+        """Checks if length of data is equal to length of column names. One set of
+        trailing commas is allowed.
+        Parameters
+        ----------
+        columns: list of column names
+        data: list of array-likes containing the data column-wise
+        """
+        if not self.index_col and len(columns) != len(data) and columns:
```

Comment:
Can you comment on why self.index_col - I find this confusing (without diving into the linked issue)

Reply:
Maybe belongs in extended_summary

Reply:
If self.index col is not False, the following returns:

```
column_names = ["one", "two", "three"]
stream = StringIO("foo,bar,baz,bam,blah")
df = pd.read_csv(stream, header=None, names=column_names, index_col=[0, 1])
```

```
         one  two three
foo bar  baz  bam  blah

```

which should not raise a Warning, hence the check

Reply:
Does this answer your question?

Reply:
Sorry - my request was not clear. I was asking for a comment to be added - which you did as part of the extended summary in the docstring. However, I do not think this is correct:

> self.index_col not False results in a ParserError previously when lengths do not match.

When index_col is True, we assume that the first N fields are part of the index, where N = difference in lengths. Isn't that what your example is showing?

Reply:
Sorry for the long wait.

I am not sure what happens if index_col=True. Per docstring this is not supported. (``index_col int, str, sequence of int / str, or False, default None``). If ``index_col=[0]`` in my example above, this raises, because lengths do not match. This is what I was referring to.

Reply:
@rhshadrach good here?

Reply:
Thanks @phofl - I think I was just confused before. doc looks great.